## BENZ PROJECT

In [1]:
#Import basic packages 
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os

In [2]:
#Read the train data 
df_train = pd.read_csv('/Users/ajithkhan/Downloads/Data Science /LMS/Machine Learning/Benz project /train.csv')
df_test = pd.read_csv('/Users/ajithkhan/Downloads/Data Science /LMS/Machine Learning/Benz project /test.csv')

In [3]:
#Lets see the shape 
print(df_train.shape)
print(df_test.shape)

(4209, 378)
(4209, 377)


In [4]:
#lets join the train and test datasets for cleaning and encoding 
df = pd.concat([df_train.assign(ind="train"), df_test.assign(ind="test")])
df.shape

(8418, 379)

##### There are 8418 samples and 379 features in this train data

In [5]:
#Lets check the concat works good or no
df.y.isnull().sum()

4209

## 1. Lets do the EDA (Checking nulls and removing low variance features)

In [6]:
df.head()

ID       y  X0 X1  X2 X3 X4 X5 X6 X8  ...  X376  X377  X378  X379  X380  \
0   0  130.81   k  v  at  a  d  u  j  o  ...     0     1     0     0     0   
1   6   88.53   k  t  av  e  d  y  l  o  ...     0     0     0     0     0   
2   7   76.26  az  w   n  c  d  x  j  x  ...     0     0     0     0     0   
3   9   80.62  az  t   n  f  d  x  l  e  ...     0     0     0     0     0   
4  13   78.02  az  v   n  f  d  h  d  n  ...     0     0     0     0     0   

   X382  X383  X384  X385    ind  
0     0     0     0     0  train  
1     0     0     0     0  train  
2     1     0     0     0  train  
3     0     0     0     0  train  
4     0     0     0     0  train  

[5 rows x 379 columns]

In [7]:
df.columns

Index(['ID', 'y', 'X0', 'X1', 'X2', 'X3', 'X4', 'X5', 'X6', 'X8',
       ...
       'X376', 'X377', 'X378', 'X379', 'X380', 'X382', 'X383', 'X384', 'X385',
       'ind'],
      dtype='object', length=379)

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8418 entries, 0 to 4208
Columns: 379 entries, ID to ind
dtypes: float64(1), int64(369), object(9)
memory usage: 24.4+ MB


#####  The available datatypes were : float64(1), int64(369), object(8)

In [9]:
col = list(df.columns)
len(col)

379

In [10]:
#Lets check the features which have more that 95% of a unique value in it.
z=[]
for x in range (2,378):
    uniq_col = df[col[x]].value_counts(normalize= True)[0] > 0.95
    if uniq_col == True:
        z.append(x)
print(z)      

[6, 10, 11, 15, 16, 17, 18, 21, 23, 24, 25, 27, 28, 29, 31, 32, 33, 35, 37, 38, 39, 40, 41, 43, 46, 47, 51, 52, 53, 54, 55, 56, 58, 59, 61, 62, 64, 65, 66, 68, 71, 73, 74, 75, 76, 77, 80, 81, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 95, 97, 100, 102, 103, 104, 105, 106, 107, 108, 110, 111, 115, 119, 120, 121, 122, 123, 127, 128, 131, 132, 135, 137, 138, 140, 142, 143, 144, 145, 148, 149, 155, 156, 158, 161, 162, 163, 165, 166, 168, 169, 170, 171, 172, 175, 179, 180, 181, 185, 187, 189, 190, 191, 192, 193, 194, 197, 198, 200, 201, 204, 205, 206, 207, 208, 210, 211, 215, 216, 220, 221, 222, 224, 225, 226, 227, 229, 230, 231, 233, 234, 236, 237, 239, 242, 243, 246, 247, 248, 249, 251, 252, 253, 254, 256, 258, 260, 261, 262, 263, 264, 265, 266, 268, 270, 271, 272, 273, 274, 275, 276, 278, 281, 282, 283, 284, 285, 286, 287, 289, 290, 291, 292, 293, 296, 298, 299, 300, 301, 302, 303, 305, 308, 310, 311, 312, 313, 315, 316, 318, 319, 321, 323, 325, 326, 328, 331, 332, 333, 334, 335, 337, 338, 

In [11]:
len(z)

233

#### So, we can neglect 233 columns out of 378 columns, since they have more than 95% of the one unique value.

In [12]:
# Lets make a copy and remove all those columns 
df1 = df.copy()
df1.head()

ID       y  X0 X1  X2 X3 X4 X5 X6 X8  ...  X376  X377  X378  X379  X380  \
0   0  130.81   k  v  at  a  d  u  j  o  ...     0     1     0     0     0   
1   6   88.53   k  t  av  e  d  y  l  o  ...     0     0     0     0     0   
2   7   76.26  az  w   n  c  d  x  j  x  ...     0     0     0     0     0   
3   9   80.62  az  t   n  f  d  x  l  e  ...     0     0     0     0     0   
4  13   78.02  az  v   n  f  d  h  d  n  ...     0     0     0     0     0   

   X382  X383  X384  X385    ind  
0     0     0     0     0  train  
1     0     0     0     0  train  
2     1     0     0     0  train  
3     0     0     0     0  train  
4     0     0     0     0  train  

[5 rows x 379 columns]

In [13]:
#Lets remove all those unnecessery columns
df2 = df1.drop(df.columns[[6, 10, 11, 15, 16, 17, 18, 21, 23, 24, 25, 27, 28, 29, 31, 32, 33, 35, 37, 38, 39, 40, 41, 43, 46, 47, 51, 52, 53, 54, 55, 56, 58, 59, 61, 62, 64, 65, 66, 68, 71, 73, 74, 75, 76, 77, 80, 81, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 95, 97, 100, 102, 103, 104, 105, 106, 107, 108, 110, 111, 115, 119, 120, 121, 122, 123, 127, 128, 131, 132, 135, 137, 138, 140, 142, 143, 144, 145, 148, 149, 155, 156, 158, 161, 162, 163, 165, 166, 168, 169, 170, 171, 172, 175, 179, 180, 181, 185, 187, 189, 190, 191, 192, 193, 194, 197, 198, 200, 201, 204, 205, 206, 207, 208, 210, 211, 215, 216, 220, 221, 222, 224, 225, 226, 227, 229, 230, 231, 233, 234, 236, 237, 239, 242, 243, 246, 247, 248, 249, 251, 252, 253, 254, 256, 258, 260, 261, 262, 263, 264, 265, 266, 268, 270, 271, 272, 273, 274, 275, 276, 278, 281, 282, 283, 284, 285, 286, 287, 289, 290, 291, 292, 293, 296, 298, 299, 300, 301, 302, 303, 305, 308, 310, 311, 312, 313, 315, 316, 318, 319, 321, 323, 325, 326, 328, 331, 332, 333, 334, 335, 337, 338, 339, 340, 342, 346, 350, 352, 357, 358, 359, 362, 363, 364, 365, 366, 371, 372, 373, 374, 375, 376, 377]],axis = 1)
df2

ID       y  X0  X1  X2 X3  X5 X6 X8  X12  ...  X361  X362  X363  X367  \
0        0  130.81   k   v  at  a   u  j  o    0  ...     1     0     0     0   
1        6   88.53   k   t  av  e   y  l  o    0  ...     1     0     1     0   
2        7   76.26  az   w   n  c   x  j  x    0  ...     1     0     1     0   
3        9   80.62  az   t   n  f   x  l  e    0  ...     1     0     1     0   
4       13   78.02  az   v   n  f   h  d  n    0  ...     1     0     1     0   
...    ...     ...  ..  ..  .. ..  .. .. ..  ...  ...   ...   ...   ...   ...   
4204  8410     NaN  aj   h  as  f  aa  j  e    0  ...     1     0     1     0   
4205  8411     NaN   t  aa  ai  d  aa  j  y    0  ...     1     1     1     0   
4206  8413     NaN   y   v  as  f  aa  d  w    0  ...     1     1     1     0   
4207  8414     NaN  ak   v  as  a  aa  c  q    0  ...     1     0     1     0   
4208  8416     NaN   t  aa  ai  c  aa  g  r    0  ...     1     1     0     0   

      X368  X374  X375  X376  X377    ind  
0        0     0     0     0     1  train  
1        0     0     1     0     0  train  
2        0     0     0     0     0  train  
3        0     0     0     0     0  train  
4        0     0     0     0     0  train  
...    ...   ...   ...   ...   ...    ...  
4204     0     1     0     0     0   test  
4205     0     0     0     1     0   test  
4206     0     1     0     0     0   test  
4207     0     0     0     0     1   test  
4208     0     0     1     0     0   test  

[8418 rows x 146 columns]

### This df2 is the final dataframe in which unnecessary features with more than 95% of single unique valve is neglected

In [14]:
#Lets cross check with any other feature 
df2.X12.value_counts(normalize= True)

0    0.925279
1    0.074721
Name: X12, dtype: float64

#### Good to go 

## 2. Applying Label Encoder (getdummies method)

In [15]:
df.X0.nunique()

53

In [16]:
X0_types = df2.X0.unique()

In [17]:
X0dum = pd.get_dummies(df['X0'], prefix_sep='_', prefix='X0')
df2 = pd.concat([df2, X0dum], axis=1)
df2.drop('X0', axis=1, inplace=True)
df2.head()

ID       y X1  X2 X3 X5 X6 X8  X12  X13  ...  X0_q  X0_r  X0_s  X0_t  X0_u  \
0   0  130.81  v  at  a  u  j  o    0    1  ...     0     0     0     0     0   
1   6   88.53  t  av  e  y  l  o    0    0  ...     0     0     0     0     0   
2   7   76.26  w   n  c  x  j  x    0    0  ...     0     0     0     0     0   
3   9   80.62  t   n  f  x  l  e    0    0  ...     0     0     0     0     0   
4  13   78.02  v   n  f  h  d  n    0    0  ...     0     0     0     0     0   

   X0_v  X0_w  X0_x  X0_y  X0_z  
0     0     0     0     0     0  
1     0     0     0     0     0  
2     0     0     0     0     0  
3     0     0     0     0     0  
4     0     0     0     0     0  

[5 rows x 198 columns]

In [18]:
X1dum = pd.get_dummies(df['X1'], prefix_sep='_', prefix='X1')
df2 = pd.concat([df2, X1dum], axis=1)
df2.drop('X1', axis=1, inplace=True)
df2.head()

ID       y  X2 X3 X5 X6 X8  X12  X13  X14  ...  X1_p  X1_q  X1_r  X1_s  \
0   0  130.81  at  a  u  j  o    0    1    0  ...     0     0     0     0   
1   6   88.53  av  e  y  l  o    0    0    0  ...     0     0     0     0   
2   7   76.26   n  c  x  j  x    0    0    0  ...     0     0     0     0   
3   9   80.62   n  f  x  l  e    0    0    0  ...     0     0     0     0   
4  13   78.02   n  f  h  d  n    0    0    0  ...     0     0     0     0   

   X1_t  X1_u  X1_v  X1_w  X1_y  X1_z  
0     0     0     1     0     0     0  
1     1     0     0     0     0     0  
2     0     0     0     1     0     0  
3     1     0     0     0     0     0  
4     0     0     1     0     0     0  

[5 rows x 224 columns]

In [19]:
X2dum = pd.get_dummies(df['X2'], prefix_sep='_', prefix='X2')
df2 = pd.concat([df2, X2dum], axis=1)
df2.drop('X2', axis=1, inplace=True)

In [20]:
X3dum = pd.get_dummies(df['X3'], prefix_sep='_', prefix='X3')
df2 = pd.concat([df2, X3dum], axis=1)
df2.drop('X3', axis=1, inplace=True)

In [21]:
X5dum = pd.get_dummies(df['X5'], prefix_sep='_', prefix='X5')
df2 = pd.concat([df2, X5dum], axis=1)
df2.drop('X5', axis=1, inplace=True)
df2.head()

ID       y X6 X8  X12  X13  X14  X19  X20  X22  ...  X5_q  X5_r  X5_s  \
0   0  130.81  j  o    0    1    0    0    0    0  ...     0     0     0   
1   6   88.53  l  o    0    0    0    0    0    0  ...     0     0     0   
2   7   76.26  j  x    0    0    0    0    0    0  ...     0     0     0   
3   9   80.62  l  e    0    0    0    0    0    0  ...     0     0     0   
4  13   78.02  d  n    0    0    0    0    0    0  ...     0     0     0   

   X5_t  X5_u  X5_v  X5_w  X5_x  X5_y  X5_z  
0     0     1     0     0     0     0     0  
1     0     0     0     0     0     1     0  
2     0     0     0     0     1     0     0  
3     0     0     0     0     1     0     0  
4     0     0     0     0     0     0     0  

[5 rows x 311 columns]

In [22]:
X6dum = pd.get_dummies(df['X6'], prefix_sep='_', prefix='X6')
df2 = pd.concat([df2, X6dum], axis=1)
df2.drop('X6', axis=1, inplace=True)

In [23]:
X8dum = pd.get_dummies(df['X8'], prefix_sep='_', prefix='X8')
df2 = pd.concat([df2, X8dum], axis=1)
df2.drop('X8', axis=1, inplace=True)

In [24]:
df2.dtypes

ID        int64
y       float64
X12       int64
X13       int64
X14       int64
         ...   
X8_u      uint8
X8_v      uint8
X8_w      uint8
X8_x      uint8
X8_y      uint8
Length: 346, dtype: object

In [51]:
df2.shape

(8418, 346)

#### This is the final dataset after encoding the Categorical values

In [25]:
# Now lets split back to the test and train data after encoding 
df2_test, df2_train = df2[df2["ind"].eq("test")], df2[df2["ind"].eq("train")]
df2_test = df2_test.drop(['y'], axis =1)
print(df2_test.shape , df2_train.shape)

(4209, 345) (4209, 346)


In [26]:
#Xtrain and ytrain
Xtrain = df2_train.drop(['y', 'ID', 'ind'], axis =1)
ytrain = df2_train['y']
Xtest = df2_test.drop(['ID', 'ind'], axis =1)

## 3. Dimensionality Reduction

In [27]:
#Lets build a base linear Regression model 
from sklearn.linear_model import LinearRegression
from sklearn.metrics import accuracy_score

In [28]:
#Fit the training set 
lr = LinearRegression()

In [29]:
lr.fit(Xtrain,ytrain)

LinearRegression()

In [30]:
#Calculate the accuracy 
lr.score(Xtrain,ytrain)

0.6045107677925108

##### The base model linear Regression scored 60%

In [31]:
#Lets reduce the dimension (PCA)
from sklearn.decomposition import PCA

In [32]:
#Fitting the data 
pca = PCA(n_components=0.95)
pca.fit(Xtrain)

PCA(n_components=0.95)

In [33]:
Xtrain_transformed = pca.transform(Xtrain)

#Lets check the reduction
print(Xtrain.shape)
print(Xtrain_transformed.shape)

In [54]:
print(Xtrain.shape)
print(Xtrain_transformed.shape)

(4209, 343)
(4209, 112)


#### So, using PCA, 231 features were reduced. 

In [34]:
#Lets re-run the Linear regression and check for any improvement
lr.fit(Xtrain_transformed, ytrain)

LinearRegression()

In [35]:
lr.score(Xtrain_transformed, ytrain)

0.5494779073043663

##### The Linear Regression model does not benefits from dimensionality reduction 

## 4. Boosting using XGBoost

In [36]:
#Import the packages
from sklearn.model_selection import StratifiedKFold
from sklearn.model_selection import cross_val_score
from xgboost import XGBRFRegressor

/Users/ajithkhan/opt/anaconda3/lib/python3.9/site-packages/xgboost/compat.py:36: FutureWarning: pandas.Int64Index is deprecated and will be removed from pandas in a future version. Use pandas.Index with the appropriate dtype instead.
  from pandas import MultiIndex, Int64Index


In [44]:
#Intiate the model and calculate the accuracy  
xgb = XGBRFRegressor()
xgb.fit(Xtrain_transformed, ytrain)
score = xgb.score(Xtrain_transformed, ytrain)
scores = cross_val_score(xgb, Xtrain_transformed, ytrain,cv=10)
print(score)
print("Mean cross-validation score: %.2f" % scores.mean())

0.6259713364833961
Mean cross-validation score: 0.48


##### The accuracy increased to 62.5% from a linear Regressor model(60%)

In [57]:
#lets predict using X_test_transformed
Xtest_transformed = pca.transform(Xtest)
y_transformed_predict = xgb.predict(Xtest_transformed)
y_transformed_predict

array([ 77.72224 ,  94.45944 ,  82.90656 , ...,  96.23722 , 111.717224,
        93.86682 ], dtype=float32)

## CONCLUSION 

#### 1. The column have been reduced from 378 feature to 146 columns by neglecting the column with almost 0% variance 

#### 2. There are no null values in the dataset 

#### 3. One hot encoding was done for the categorical features and the number of features increased from 146 to 346

#### 4.Done PCA to reduce the dimension from 346 to 112

#### 5. Prediction was done in X_test_transformed using XGBoost and the results were stored in y_transformed_predict